### PR TITLE
ADIOS2: Optionally write attributes only from given ranks

### PR DESCRIPTION
**Pull request by franzpoeschel**
Performance optimization especially for large scale
Ref.
https://github.com/ornladios/ADIOS2/issues/3846#issuecomment-1769255906

TODO:

- [x] Documentation
- [ ] Is this relevant for HDF5 too?

### CODE REVIEW

File: docs/source/backends/adios2.rst [hunk 1]
```diff
@@ -135,6 +140,24 @@ The preferred backend usually depends on the system's native software stack.
...
+The ADIOS2 backend of the openPMD-api will then ignore attributes from all other MPI ranks.
+
+.. tip::
+
+  Treat metadata specification as a collective operation in order to retain compatibility with HDF5, and then specify ``adios2.attribute_writing_ranks = 0`` in order to achieve best performance in ADIOS2.
+
+.. warning::
+
+  The ADIOS2 backend may also use attributes to encode openPMD groups (ref. "group table").
+  The ``adios.attribute_writing_ranks`` key also applies to those attributes, i.e. also group creation must be treated as collective then (at least on the specified ranks).
```
Comment (ax3l) [inline docs/source/backends/adios2.rst]:
```suggestion
  The ``adios2.attribute_writing_ranks`` key also applies to those attributes, i.e. also group creation must be treated as collective then (at least on the specified ranks).
```

**Review by ax3l — APPROVED** [review]

**Comment by ax3l** [comment]
@eschnett thinking about portable code that also supports HDF5, would this interrupt you if we enabled this by default (in a follow-up PR)? :sweat_smile:

**Comment by eschnett** [comment]
@ax3l I don't understand the ramifications. On the surface this all seems benign and beneficial.

**Comment by franzpoeschel** [comment]
@eschnett The possible pitfall is that this would by default ignore any attribute that is not written on rank 0.
Background:

* attribute writes in HDF5 are collective
* attribute writes in ADIOS2 are non-collective, and defining them only on one rank is more performant than doing it collectively

In consequence: Portable code using the openPMD-api should treat attribute writes as something collective. If attributes are written collectively, there's no harm in ADIOS2 to ignore any attribute write except for those on rank 0. This is why we're thinking about enabling this by default.

If for some reason you have routines that create some data only on a few select ranks, this could lead to problems.

**Comment by franzpoeschel** [comment]
@ax3l If we decide to activate this by default, we should maybe add some easy way to activate attribute writing from all ranks. Currently, `attribute_writing_ranks` is a list of ranks that opt into writing attributes. Something like `attribute_writing_ranks = "all"` might then be needed to restore the current behavior.

**Comment by eschnett** [comment]
Thanks for the explanation.

I was not aware that openPMD treated attribute writes as global. This will be good to know.

**Pull request merged**